File: bot-core/src/main/kotlin/fr/delphes/dynamicForm/descriptor/FormListDescriptor.kt
package fr.delphes.dynamicForm.descriptor

import fr.delphes.annotation.dynamicForm.DynamicFormFamily
import fr.delphes.dynamicForm.DynamicFormDTO
import kotlinx.serialization.Contextual
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@SerialName("FORM_LIST")
@Serializable
data class FormListFieldDescriptor(
    override val fieldName: String,
    override val description: String,
    val formFamily: DynamicFormFamily,
    override val value: List<DynamicFormDTO<@Contextual Any>>,
) : FieldDescriptor()

File: bot-ui/src/common/dynamicForm/field/formList/form-list-descriptor.ts
import {DynamicFormFamily} from "@/common/dynamicForm/dynamic-form-family";
import {FieldDescriptor, FieldValue} from "@/common/dynamicForm/field/field-descriptor";
import {FieldDescriptorType} from "@/common/dynamicForm/field/field-descriptor-type";
import {DescriptorJsonType} from "@/common/dynamicForm/field/field-descriptor.factory";
import FieldFormListEditView from "@/common/dynamicForm/field/formList/field-form-list-edit-view.vue";

export class FormListDescriptor implements FieldDescriptor<string> {
    static readonly type: FieldDescriptorType = FieldDescriptorType.FORM_LIST;

    constructor(
        readonly description: string,
        readonly fieldName: string,
        readonly formFamily: DynamicFormFamily,
        readonly initialValue: string,
        readonly actualValue: string = initialValue
    ) {
    }

    static fromJson(json: DescriptorJsonType): FormListDescriptor {
        if (json.type !== this.type) {
            throw new Error(`Cannot deserialize ${json.type} as FormListDescriptor`);
        }

        return new FormListDescriptor(
            json.description,
            json.fieldName,
            json.formFamily,
            JSON.stringify(json.value)
        );
    }

    buildValue() {
        return new FieldValue(this.fieldName, JSON.parse(this.actualValue));
    }

    viewComponent() {
        return FieldFormListEditView
    }
}